src/setup.rs

```diff
@@ -207,24 +233,104 @@ pub fn setup(
     )
 }
 
-pub struct AllIOs {
-    pub io: Pin<Gpio10, PushPullOutput>,
-    pub ck: Pin<Gpio11, PushPullOutput>,
-    pub tdi: Pin<Gpio17, PushPullOutput>,
-    pub tdo_swo: Pin<Gpio16, PushPullOutput>,
-    pub reset: Pin<Gpio9, PushPullOutput>,
-    pub vcp_rx: Pin<Gpio21, PushPullOutput>,
-    pub vcp_tx: Pin<Gpio20, PushPullOutput>,
+/// A simple Dynamic pin, meaning it can be output or input.
+pub enum DynPin<P, InputPull>
+where
+    P: PinId,
+    InputPull: PullType,
+{
+    /// The pin is in input mode.
+    Input(Pin<P, FunctionSioInput, InputPull>),
+    /// The pin is in output mode.
+    Output(Pin<P, FunctionSioOutput, InputPull>),
+}
+
+impl<P, InputPull> DynPin<P, InputPull>
+where
+    P: ValidFunction<FunctionSioInput>,
+    P: ValidFunction<FunctionSioOutput>,
+    InputPull: PullType,
+{
+    pub fn into_input(&mut self) {
+        unsafe {
+            replace_with_or_abort_unchecked(self, |p| match p {
+                DynPin::Input(_) => p,
+                DynPin::Output(o) => DynPin::Input(o.into_function().into_pull_type()),
+            })
+        }
```

Comment:
For these I would rather use an option instead because it avoids unsafe code.  So hold the enumeration in an option, then you can use _take_ to move the pin from a mutable borrow of self.

Reply:
While it removes the need for unsafe you loose the compile time checking of state transistions, so I'm not sure it's better. One could just wrap the replace, and only have a single `unsafe`. It's only unsafe if you have unwinding, which we don't.

Reply:
Happy to learn something new!  How does your code provide compile time state transition checking?

Reply:
The pattern of `replace_with` forces you to put back an object of the same type - so e.g. getting stuck in the `None` case, if you use `Option<Thing>`, is impossible by construction.
It's a property I quite enjoy as the `None` case have bitten me many times before. :)

Reply:
I see what you're saying now, I thought you were telling me the enum values were checked at compile time somehow.